lib/selinux/selinux_linux.go

```diff
@@ -0,0 +1,177 @@
+	moduleName           = "teleport_ssh"
+	domain               = "teleport_ssh_t"
+	permissiveModuleName = "permissive_" + domain
+)
+
+//go:embed teleport_ssh.te
+var module string
+
+//go:embed teleport_ssh.fc.tmpl
+var fileContexts string
+
+// ModuleSource returns the source of the SELinux SSH module.
+func ModuleSource() string {
+	return module
+}
+
+type filePaths struct {
+	InstallDir     string
+	DataDir        string
+	ConfigPath     string
+	UpgradeUnitDir string
+}
+
+// FileContexts returns file contexts for the SELinux SSH module.
+func FileContexts(installDir, dataDir, configPath string) (string, error) {
+	fcTempl, err := template.New("selinux file contexts").Parse(fileContexts)
+	if err != nil {
+		return "", trace.Wrap(err, "failed to parse file contexts template")
+	}
+
+	// Generate a file specifying the locations of important dirs so SELinux
+	// will allow Teleport SSH to be able to access them.
+	var buf bytes.Buffer
+	err = fcTempl.Execute(&buf, filePaths{
+		InstallDir:     installDir,
+		DataDir:        dataDir,
+		ConfigPath:     configPath,
+		UpgradeUnitDir: versioncontrol.UnitConfigDir,
+	})
+	if err != nil {
+		return "", trace.Wrap(err, "failed to expand file contexts template")
+	}
+
+	return buf.String(), nil
+}
+
+// CheckConfiguration returns an error if SELinux is not configured to
+// enforce the SSH service correctly.
+func CheckConfiguration(ensureEnforced bool, logger *slog.Logger) error {
+	if !ocselinux.GetEnabled() {
+		return trace.Errorf("SELinux is disabled or not present")
+	}
+	if ocselinux.EnforceMode() == ocselinux.Disabled {
+		return trace.Errorf("SELinux mode is disabled, SELinux will not enforce or log anything")
+	}
+	if ocselinux.EnforceMode() == ocselinux.Permissive {
+		if ensureEnforced {
+			return trace.Errorf("SELinux mode is permissive, SELinux will not enforce rules only log denials")
+		} else {
+			slog.WarnContext(context.TODO(), "SELinux mode is permissive, SELinux will not enforce rules only log denials")
+		}
+	}
+
+	selinuxType, err := readConfig(selinuxTypeTag)
+	if err != nil {
+		return trace.Wrap(err, "failed to find SELinux type")
+	}
+	selinuxDir := filepath.Join("/var/lib/selinux", selinuxType, "active/modules")
+
+	var moduleInstalled, moduleDisabled, modulePermissive bool
+	err = filepath.WalkDir(selinuxDir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return trace.Wrap(err, "failed to access %q", path)
+		}
+
+		name := d.Name()
+		if !strings.Contains(name, moduleName) {
+			return nil
+		}
+
+		if d.IsDir() {
+			moduleInstalled = true
+			if name == permissiveModuleName {
+				modulePermissive = true
+				// if the module is permissive, we also know it's installed
+				// so we can end the walk
+				return filepath.SkipAll
+			}
+			// if the module is disabled, an empty file with the module's
+			// name will exist in the "disabled" dir
+		} else if filepath.Base(filepath.Dir(path)) == "disabled" {
```

Comment:
Can the `disabled` dir be at any level relative to the module? If not, can we just check if `<module>/../disabled/<module>` exists?

Reply:
done